File: tests/fixtures/go-gin/handler/user_handler.go
// handler/user_handler.go - User HTTP handlers

package handler

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"

	"myapp/models"
	"myapp/repository"
	"myapp/service"
)

type UserHandler struct {
	userService *service.UserService
}

func NewUserHandler(userService *service.UserService) *UserHandler {
	return &UserHandler{userService: userService}
}

// RegisterRoutes registers user routes
func (h *UserHandler) RegisterRoutes(r *gin.RouterGroup) {
	users := r.Group("/users")
	{
		users.GET("", h.ListUsers)
		users.GET("/:id", h.GetUser)
		users.POST("", h.CreateUser)
		users.PUT("/:id", h.UpdateUser)
		users.DELETE("/:id", h.DeleteUser)
		users.GET("/me", h.GetCurrentUser)
		users.PATCH("/me/profile", h.UpdateProfile)
		users.PATCH("/me/preferences", h.UpdatePreferences)
		users.POST("/me/change-password", h.ChangePassword)
		users.POST("/:id/activate", h.ActivateUser)
		users.POST("/:id/suspend", h.SuspendUser)
		users.GET("/stats", h.GetStats)
		users.GET("/search", h.SearchUsers)
	}
}

func (h *UserHandler) ListUsers(c *gin.Context) {
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "20"))

	opts := repository.ListOptions{
		Page:     page,
		PageSize: pageSize,
		Role:     c.Query("role"),
		Status:   c.Query("status"),
		Search:   c.Query("search"),
	}

	users, total, err := h.userService.ListUsers(c.Request.Context(), opts)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"users":       users,
		"total":       total,
		"page":        page,
		"page_size":   pageSize,
		"total_pages": (total + int64(pageSize) - 1) / int64(pageSize),
	})
}

func (h *UserHandler) GetUser(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid user ID"})
		return
	}

	user, err := h.userService.GetUser(c.Request.Context(), id)
	if err != nil {
		if err == service.ErrUserNotFound {
			c.JSON(http.StatusNotFound, gin.H{"error": "user not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, user)
}

// Q: How should we handle input validation and return structured error responses?
func (h *UserHandler) CreateUser(c *gin.Context) {
	var input service.CreateUserInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	input.SendVerification = true

	user, err := h.userService.CreateUser(c.Request.Context(), input)
	if err != nil {
		switch err {
		case service.ErrEmailExists:
			c.JSON(http.StatusConflict, gin.H{"error": "email already registered"})
		case service.ErrUsernameExists:
			c.JSON(http.StatusConflict, gin.H{"error": "username already taken"})
		default:
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		}
		return
	}

	c.JSON(http.StatusCreated, user)
}

func (h *UserHandler) UpdateUser(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid user ID"})
		return
	}

	var updates map[string]interface{}
	if err := c.ShouldBindJSON(&updates); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	user, err := h.userService.UpdateUser(c.Request.Context(), id, updates)
	if err != nil {
		if err == service.ErrUserNotFound {
			c.JSON(http.StatusNotFound, gin.H{"error": "user not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, user)
}

func (h *UserHandler) DeleteUser(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid user ID"})
		return
	}

	if err := h.userService.DeleteUser(c.Request.Context(), id); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

func (h *UserHandler) GetCurrentUser(c *gin.Context) {
	userID, _ := c.Get("user_id")
	id := userID.(uuid.UUID)

	user, err := h.userService.GetUser(c.Request.Context(), id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, user)
}

func (h *UserHandler) UpdateProfile(c *gin.Context) {
	userID, _ := c.Get("user_id")
	id := userID.(uuid.UUID)

	var input service.UpdateProfileInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	profile, err := h.userService.UpdateProfile(c.Request.Context(), id, input)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, profile)
}

func (h *UserHandler) UpdatePreferences(c *gin.Context) {
	userID, _ := c.Get("user_id")
	id := userID.(uuid.UUID)

	// Would implement preferences update
	_ = id

	c.JSON(http.StatusOK, gin.H{"message": "preferences updated"})
}

func (h *UserHandler) ChangePassword(c *gin.Context) {
	userID, _ := c.Get("user_id")
	id := userID.(uuid.UUID)

	var input struct {
		CurrentPassword string `json:"current_password" binding:"required"`
		NewPassword     string `json:"new_password" binding:"required,min=8"`
	}

	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if err := h.userService.ChangePassword(c.Request.Context(), id, input.CurrentPassword, input.NewPassword); err != nil {
		if err == service.ErrInvalidPassword {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "incorrect current password"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "password changed successfully"})
}

func (h *UserHandler) ActivateUser(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid user ID"})
		return
	}

	user, err := h.userService.ActivateUser(c.Request.Context(), id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, user)
}

func (h *UserHandler) SuspendUser(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid user ID"})
		return
	}

	var input struct {
		Reason       string `json:"reason"`
		DurationDays int    `json:"duration_days"`
	}
	c.ShouldBindJSON(&input)

	user, err := h.userService.SuspendUser(c.Request.Context(), id, input.Reason, input.DurationDays)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, user)
}

func (h *UserHandler) GetStats(c *gin.Context) {
	stats, err := h.userService.GetStats(c.Request.Context())
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, stats)
}

func (h *UserHandler) SearchUsers(c *gin.Context) {
	query := c.Query("q")
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "20"))

	users, err := h.userService.SearchUsers(c.Request.Context(), query, limit)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, users)
}

File: tests/fixtures/rust-axum/src/handlers/user_handler.rs
// src/handlers/user_handler.rs - User HTTP handlers

use std::sync::Arc;

use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::error::AppError;
use crate::models::user::{UserResponse, UserRole, UserStatus};
use crate::repository::user_repository::ListOptions;
use crate::services::user_service::{CreateUserInput, UpdateUserInput, UserService};
use crate::AppState;

pub async fn list_users(
    State(state): State<Arc<AppState>>,
    Query(params): Query<ListUsersParams>,
) -> Result<impl IntoResponse, AppError> {
    let opts = ListOptions {
        page: params.page.unwrap_or(1),
        page_size: params.page_size.unwrap_or(20),
        role: params.role,
        status: params.status,
        search: params.search,
    };

    let (users, total) = state.user_service.list_users(opts).await?;
    let page = params.page.unwrap_or(1);
    let page_size = params.page_size.unwrap_or(20);

    Ok(Json(ListUsersResponse {
        users: users.into_iter().map(|u| UserResponse::from((u, None))).collect(),
        total,
        page,
        page_size,
        total_pages: (total as f64 / page_size as f64).ceil() as i64,
    }))
}

pub async fn get_user(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    let user = state.user_service.get_user(id).await?;
    Ok(Json(UserResponse::from((user, None))))
}

// Q: How should we structure error responses for better client-side handling?
pub async fn create_user(
    State(state): State<Arc<AppState>>,
    Json(input): Json<CreateUserRequest>,
) -> Result<impl IntoResponse, AppError> {
    let user = state
        .user_service
        .create_user(CreateUserInput {
            email: input.email,
            username: input.username,
            password: input.password,
            first_name: input.first_name,
            last_name: input.last_name,
            send_verification: true,
        })
        .await?;

    Ok((StatusCode::CREATED, Json(UserResponse::from((user, None)))))
}

pub async fn update_user(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
    Json(input): Json<UpdateUserRequest>,
) -> Result<impl IntoResponse, AppError> {
    let user = state
        .user_service
        .update_user(
            id,
            UpdateUserInput {
                username: input.username,
                role: input.role,
                status: input.status,
            },
        )
        .await?;

    Ok(Json(UserResponse::from((user, None))))
}

pub async fn delete_user(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    state.user_service.delete_user(id).await?;
    Ok(StatusCode::NO_CONTENT)
}

pub async fn get_current_user(
    State(state): State<Arc<AppState>>,
    // Would extract user_id from auth middleware
) -> Result<impl IntoResponse, AppError> {
    // Placeholder - would get user_id from auth context
    let user_id = Uuid::nil();
    let user = state.user_service.get_user(user_id).await?;
    Ok(Json(UserResponse::from((user, None))))
}

pub async fn change_password(
    State(state): State<Arc<AppState>>,
    Json(input): Json<ChangePasswordRequest>,
) -> Result<impl IntoResponse, AppError> {
    // Would get user_id from auth context
    let user_id = Uuid::nil();
    state
        .user_service
        .change_password(user_id, &input.current_password, &input.new_password)
        .await?;

    Ok(Json(MessageResponse {
        message: "Password changed successfully".to_string(),
    }))
}

pub async fn activate_user(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    let user = state.user_service.activate_user(id).await?;
    Ok(Json(UserResponse::from((user, None))))
}

pub async fn suspend_user(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
    Json(input): Json<SuspendUserRequest>,
) -> Result<impl IntoResponse, AppError> {
    let user = state
        .user_service
        .suspend_user(id, &input.reason, input.duration_days)
        .await?;

    Ok(Json(UserResponse::from((user, None))))
}

pub async fn get_stats(
    State(state): State<Arc<AppState>>,
) -> Result<impl IntoResponse, AppError> {
    let stats = state.user_service.get_stats().await?;
    Ok(Json(stats))
}

pub async fn search_users(
    State(state): State<Arc<AppState>>,
    Query(params): Query<SearchUsersParams>,
) -> Result<impl IntoResponse, AppError> {
    let users = state
        .user_service
        .search_users(&params.q, params.limit.unwrap_or(20))
        .await?;

    Ok(Json(users.into_iter().map(|u| UserResponse::from((u, None))).collect::<Vec<_>>()))
}

// Request/Response types
#[derive(Debug, Deserialize)]
pub struct ListUsersParams {
    pub page: Option<i32>,
    pub page_size: Option<i32>,
    pub role: Option<String>,
    pub status: Option<String>,
    pub search: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ListUsersResponse {
    pub users: Vec<UserResponse>,
    pub total: i64,
    pub page: i32,
    pub page_size: i32,
    pub total_pages: i64,
}

#[derive(Debug, Deserialize)]
pub struct CreateUserRequest {
    pub email: String,
    pub username: String,
    pub password: String,
    pub first_name: Option<String>,
    pub last_name: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateUserRequest {
    pub username: Option<String>,
    pub role: Option<UserRole>,
    pub status: Option<UserStatus>,
}

#[derive(Debug, Deserialize)]
pub struct ChangePasswordRequest {
    pub current_password: String,
    pub new_password: String,
}

#[derive(Debug, Deserialize)]
pub struct SuspendUserRequest {
    pub reason: String,
    pub duration_days: Option<i32>,
}

#[derive(Debug, Deserialize)]
pub struct SearchUsersParams {
    pub q: String,
    pub limit: Option<i32>,
}

#[derive(Debug, Serialize)]
pub struct MessageResponse {
    pub message: String,
}
